### PR TITLE
Expand Stripe Webhook Integration

### PR DESCRIPTION
**Pull request by cadecairos**
I'll be adding in support for more stripe webhook events and the ability to pass the data to the SQS queue that basket consumes and syncs with Salesforce.

### CODE REVIEW

**Comment by cadecairos** [comment]
Here's some examples of different scenarios and what they'll dump into the queue:

### charge.failed (network declined)

```
{ timestamp: '2017-07-19T16:40:04.131Z',
  data:
   { event_type: "charge.failed",
     transaction_id: 'ch_1AhLDUG8Mmx3htnx1dhlWjfW',
     failure_code: 'card_declined',
     failure_message: 'Your card was declined.',
     outcome:
      { network_status: 'declined_by_network',
        reason: 'generic_decline',
        risk_level: 'normal',
        seller_message: 'The bank did not return any further details with this decline.',
        type: 'issuer_declined' } } }
```

### charge.refunded (marked fraudulent by mofo/stripe)

#### Full refund (Fraudulent)

```
{ timestamp: '2017-07-19T16:34:23.622Z',
  data:
   { event_type: "charge.refunded",
     transaction_id: 'ch_1AhKbGG8Mmx3htnxvHJUun42',
     amount_refunded: 20.00,
     currency: "USD",
     fully_refunded: true,
     reason: 'fraudulent' } }
```

#### Partial refund (made through dashboard - eg. by support specialist)

```
{ timestamp: '2017-07-19T18:11:12.581Z',
  data:
   { event.type: "charge.refunded",
     transaction_id: 'ch_1AbuggG8Mmx3htnxCJoUpWzb',
     amount_refunded: 2.50,
     currency: "USD",
     fully_refunded: false,
     reason: 'donor request' } }
```

### charge.dispute.closed | charge.dispute.created | charge.dispute.funds_reinstated

```
{ timestamp: '2017-07-19T16:57:06.023Z',
  data:
   { event.type: "charge.dispute.closed",
     transaction_id: 'ch_1AhLTgG8Mmx3htnx4SWwMyai',
     dispute_amount: 400.00,
     dispute_currency: "USD",
     disputed_on: 1500483409,
     dispute_currency: 'usd',
     dispute_reason: 'fraudulent',
     dispute_status: 'lost' } }
```

Note: the dispute reason and status will change depending on the event and the data associated with it.

**Comment by cadecairos** [comment]
My comment above is out of date - I slimmed down everything a bit.

based on the documentation provided by Desiree, Stephanie, and Vojtech, here's what this code does:

#### If a charge fails
```
{
  "event_type": "charge.failed",
  "transaction_id": "[charge ID to look up in Salesforce]",
  "failure_code": "[failure code from payment network - i.e. expired_card]"
}
```

#### a charge is refunded
```
{
  "event_type": "charge.refunded",
  "transaction_id": "[charge ID to look up in Salesforce]",
  "reason": "[reason for refund - i.e 'requested_by_customer'",
  "status": "[status of refund - i.e. successful]"
```

#### a dispute is created, updated, or closed
```
{
  "event_type": "charge.dispute.[created|updated|closed]",
  "transaction_id": "[charge ID to look up in Salesforce]",
  "reason": "reason given for dispute i.e 'fraudulent'",
  "status": "state of dispute, i.e. 'lost'" 
}
```

File: routes/index.js [hunk 1]
```diff
@@ -467,23 +467,104 @@ var routes = {
...
+    var transaction_id = charge.id;
+    var reason = refund.reason;
+    var status = refund.status;
+
+    if (reason === null) {
+      // refunded via dashboard, mark as requested_by_customer
+      reason = 'requested_by_customer';
+    }
+
+    basket.queue({
```
Comment (pmac) [inline routes/index.js]: This one needs property names. Like this it's just an array I think?

Comment (cadecairos) [inline routes/index.js]: This is JS object notation shorthand, the property names will be the name of the variables, and the values will be the values of the variables. 

more details here: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Object_initializer

Comment (pmac) [inline routes/index.js]: Oh that is nice! That'd have saved me soooo many keystrokes back in the day :)

**Comment by pmac** [comment]
@cadecairos are you planning on adding `event_type` to the initial donation messages as well, or should I expect those to remain as-is? I'm thinking I could potentially expect `event_type` for all messages, and if it doesn't have one assume it's a `donation` type. That seems like a good way to support the current and future states.

**Comment by cadecairos** [comment]
> @cadecairos are you planning on adding event_type to the initial donation messages as well, or should I expect those to remain as-is? I'm thinking I could potentially expect event_type for all messages, and if it doesn't have one assume it's a donation type. That seems like a good way to support the current and future states.

Yes, each message will have an event_type attribute, that mirrors the event type that triggered the Stripe webhook.

this is an example of a complete message (it's json encoded as always): 

```json
{ 
  "timestamp": "2017-07-19T16:34:23.622Z",
  "data": {
    "event_type": "charge.refunded",
    "transaction_id": "[charge ID to look up in Salesforce]",
    "reason": "[reason for refund - i.e 'requested_by_customer'",
    "status": "[status of refund - i.e. successful]"
  } 
}
```

**Comment by pmac** [comment]
Yes. I see that it'll be there for all of the new webhook messages, but was wondering about the donation messages we're currently processing. You can add `event_type` to those as well if you'd like to keep things consistent, or I can just assume that if it has no `event_type` that it's a donation.

**Comment by cadecairos** [comment]
I see what you mean now. it's probably better if I add in the event there for consistency

**Comment by cadecairos** [comment]
@ScottDowne I added a new commit that adds an additional attribute for the stripe and paypal messages we send into the queue to indicate they're donations.

Can you green light a merge?

**Comment by ScottDowne** [comment]
🍏 📗 💚 🥗 💡

**Pull request merged**